bluetooth: BLE GATT battery service

lemrey opened the pull request:
I am exploring the requirements for moving GATT services out of /samples.
The goal is to have a base of re-usable service implementations that can be easily pulled in by the application via Kconfig and that require the least amount of interfacing with the application.

- Added a mockup battery sensor driver that simulates a state of charge
- Added a Kconfig entry to enable the BLE GATT Battery service and moved it to /subsys/bluetooth/services from /samples/bluetooth/gatt
- Updated the service implementation to use the sensor interface

These changes let the user enable the service and select a data source (battery sensor) via Kconfig. No changes to the main program are required to switch data sources, making it easy to switch to real hardware.

Previous comments and PR: https://github.com/zephyrproject-rtos/zephyr/pull/9712

lemrey committed:
sensor: add state of charge channel

This commit adds a state of charge channel to measure a battery's
charge state.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey committed:
drivers: sensors: add mockup battery state of charge sensor driver

This commit adds a mockup battery state of charge sensor driver.
The driver can be used to simulate a change in battery charge level.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>

lemrey commented:
CC @carlescufi, @kapi-no, @pdunaj any feedback is appreciated.

JoeHut commented on subsys/bluetooth/services/Kconfig:
```diff
@@ -0,0 +1,33 @@
+
+config BT_GATT_BATTERY_SERVICE_MOCKUP_SENSOR
+	bool "Mockup battery sensor"
+	select MOCKUP_BATTERY_SENSOR
```
Would it make more sense to just give the device name of the sensor here?

JoeHut commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,113 @@
+#include <sensor.h>
+
+
+static struct bt_gatt_service bas_svc;
```
bas has the service already in it, right?

JoeHut commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,113 @@
+
+static struct device *battery;
+static struct sensor_value blvl = {0};
+static bool simulate_blvl;
```
``notification_enabled`` would make more sense I think, since it is no simulation but potentially a real implementation.

JoeHut commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,113 @@
+
+static void blvl_ccc_cfg_changed(const struct bt_gatt_attr *attr, u16_t value)
+{
+	simulate_blvl = (value == BT_GATT_CCC_NOTIFY) ? true : false;
```
The ``? true : false`` is redundant to the actual comparison

JoeHut commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,113 @@
+	BT_GATT_CCC(blvl_ccc_cfg, blvl_ccc_cfg_changed),
+};
+
+static struct bt_gatt_service bas_svc = BT_GATT_SERVICE(attrs);
```
Would it make sense to use ``SYS_INIT`` to automatically initialize the service similar to how it is done in the mgmt subsystem?

lemrey replied:
That's probably a good idea!

JoeHut replied:
ah, you're right. I don't know whether GATT and GAP have to be first in the list of registered services.

JoeHut reviewed (commented):
I was working on a similar implementation, so I thought my feedback could be helpful. One issue I see is that the application still has to call ``bas_init`` which could be done automatically with ``SYS_INIT`` (See https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/mgmt/smp_bt.c#L189). Also the application has to take care of doing the notifications. If the fuel gauge driver implements a DATA_READY trigger, the battery service should probably register for that callback, so that the SoC is updated without application interaction.

Another problem is, that the battery service is registered now, but it is not part of the advertisement data. This should be documented at least (I didn't find a proper way to have a distributed way of registering advertisement data).

Our current solution is more similar to what a battery subsystem would look like. It handles a fuel gauge sensor and a charging sensor so that we can process changes in the voltage, the SoC and the charging state and also can read data like the State of Health. This battery subsystem registers the triggers and the BLE battery service actually just registers a callback in the battery subsystem if notifications are enabled. This way the application can register multiple callbacks (for example to provide feedback if the battery goes below a defined threshold) and the BLE service can use a callback as well without having to handle the sensor directly.

lemrey commented:
@JoeHut thank you for you comments.

> Also the application has to take care of doing the notifications. If the fuel gauge driver implements a DATA_READY trigger, the battery service should probably register for that callback, so that the SoC is updated without application interaction.

I don't think the service should send them automatically; that's really application dependant behavior.

> Another problem is, that the battery service is registered now, but it is not part of the advertisement data. 

I see, I'll see what can be done about that, however I think this is not an issue for the battery service, since nobody would connect to a device based on whether it's advertising this service specifically.

JoeHut commented:
@lemrey 

> I see, I'll see what can be done about that, however I think this is not an issue for the battery service, since nobody would connect to a device based on whether it's advertising this service specifically.

Actually there are use cases like this :) If you advertise a BLE HID keyboard, you have to advertise a battery service and a device information service per specification. If those are not advertised, the keyboard won't show up in iOS for example.

Vudentz commented:
@JoeHut 

> Actually there are use cases like this :) If you advertise a BLE HID keyboard, you have to advertise a battery service and a device information service per specification. If those are not advertised, the keyboard won't show up in iOS for example.

I think this is specific to iOS, afaik HID includes devices info and battery service so I wonder why one would have to advertise them as well? Anyway the advertisement part is something that application will need to figure out what to include since the space available is quite tight, specially if the idea is to include every service on the GATT database.

@lemrey 

>  don't think the service should send them automatically; that's really application dependant behavior.

Im not following you here, why does the battery state has to depend on application behavior? I thought the whole point is to not have the application to implement any part of the battery service, it just selects the right Kconfig option and gets the service enabled with full functionality.

Vudentz commented on samples/bluetooth/peripheral/CMakeLists.txt:
```diff
@@ -6,7 +6,7 @@ target_sources(app PRIVATE
   ../gatt/hrs.c
   ../gatt/dis.c
-  ../gatt/bas.c
+  $ENV{ZEPHYR_BASE}/subsys/bluetooth/services/bas.c
```
This shouldn't be necessary, instead the application should just select the Kconfig option to enable BAS so in this case it would be part of prj.conf.

lemrey replied:
I created a commit where I first moved the service, then another where I created the Kconfig entry and updated the service to use the sensor interface.

carlescufi replied:
@lemrey can you please squash the 2 commits so we don't see this?

Vudentz commented on samples/bluetooth/peripheral_csc/CMakeLists.txt:
```diff
@@ -6,7 +6,7 @@ FILE(GLOB app_sources src/*.c)
   ${app_sources}
   ../gatt/dis.c
-  ../gatt/bas.c
+  $ENV{ZEPHYR_BASE}/subsys/bluetooth/services/bas.c
```
Ditto.

Vudentz commented on samples/bluetooth/peripheral/src/main.c:
```diff
@@ -22,9 +22,9 @@
 #include <bluetooth/uuid.h>
 #include <bluetooth/gatt.h>
 
+#include <bluetooth/services/bas.h>
```
This should also not be necessary, once the Kconfig option is set the service should be registered by stack itself.

lemrey commented:
@Vudentz The battery state does not depend on application behavior, but how often the battery state is notified to the peer does. I guess we could give the user the possibility to select a few options via Kconfig e.g. notify on change, notify on treshold. Still, the application might want to notify periodically.
I see two issues with this approach:
- it is hard, if possible at all, to cover all the possible 'policies' via Kconfig (on change, periodic etc.)
- the logic for all these possible 'policies' would be implemented in the battery service, this might overcomplicate things just a bit too much

JoeHut commented:
@Vudentz 
The HID over GATT profile contains the HID service, the battery service and the device information service. The HID service itself does not contain those characteristics (https://www.bluetooth.com/specifications/gatt/viewer?attributeXmlFile=org.bluetooth.service.human_interface_device.xml)

But you are right, the application should take care of that, I would just recommend to document it somewhere.

@lemrey I don't really see a reason to notify other than on value change. As long as the client gets a valid value on reading the characteristic it should be fine.

lemrey commented:
>  I don't really see a reason to notify other than on value change.

I am not convinced we can make this kind of assumptions. It could be possible, for example, that the peer can't display the level of charge and instead only turns on a LED when the battery is below a certain level. In that case it would be a waste of resources to send notifications on battery level changes.

JoeHut commented:
Yeah, you're probably right. I would just prefer something that I turn on in the config and don't have to worry about. But probably that makes too many assumptions.

Vudentz commented:
@lemrey 

> it is hard, if possible at all, to cover all the possible 'policies' via Kconfig (on change, periodic etc.) the logic for all these possible 'policies' would be implemented in the battery service, this might overcomplicate things just a bit too much

I thought such policies would be implemented by the sensor driver directly, or at least be able to select the policy and not leave it up for the application since then each and every application may come with its own specific policies which would have to be qualified separately. Also the further these policies are implemented from the driver the more APIs to expose to the application we will need to have.

lemrey commented:
@JoeHut , @Vudentz I have updated the PR based on your input.
Now the sensor driver periodically fires a callback that triggers the battery service to send notifications; 
there is Kconfig entry to set the period. Also, I have tried using SYS_INIT to initialize the service, but it seems that registering a BLE service before calling bt_enable() results in a usage fault, so only basic initialization is done in bas_init(), the service has to be registered manually using bas_register().

JoeHut commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,142 @@
+	ARG_UNUSED(attr);
+
+	notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+	printk("Battery notifications %s\n",
```
Why are you using printk instead of SYS_LOG or LOG?

JoeHut commented:
The policy is not yet used in the actual code. Generally, I think it would be helpful to have a proper fuel gauge driver in-tree. I can try to clean up and provide one for the TI BQ27421 if I find the time the next weeks. Than there would be everything working together.

lemrey commented:
Hi @JoeHut , what do you mean by "the policy is not yet used in the actual code" ?

JoeHut commented:
@lemrey Forget it. I got confused by the periodic policy. I somehow thought this would be part of the service, but it is actually in the mocked driver. I like this implementation definitely more :)

lemrey commented:
@JoeHut , correct the policy is implemented in the driver.
Also, I switched out printk in favor of the logger (in the service) and pushed that in a separate commit.

@Vudentz After you sent a PR to fix service registration I have updated this pull request so that the service is registered automatically. Now it's possible to enabled/disable this service via Kconfig exclusively, in fact main.c is not even including the header for the service any longer :+1: I'll rebase once your PR is merged.

JoeHut commented on subsys/bluetooth/services/Kconfig.bas:
```diff
@@ -0,0 +1,50 @@
+	prompt "Battery services logging level"
+	default BT_GATT_BAS_LOG_LEVEL_INF
+
+ config BT_GATT_BAS_LOG_LEVEL_OFF
```
space in front of config

JoeHut commented on subsys/bluetooth/services/Kconfig.bas:
```diff
@@ -0,0 +1,50 @@
+config BT_GATT_BAS_LOG_LEVEL_DBG
+	bool "Debug"
+
+ endchoice
```
space in front

JoeHut commented on subsys/bluetooth/services/Kconfig.bas:
```diff
@@ -0,0 +1,50 @@
+
+ endchoice
+
+ config BT_GATT_BAS_LOG_LEVEL
```
space in front of config

JoeHut commented on subsys/bluetooth/services/Kconfig.bas:
```diff
@@ -0,0 +1,50 @@
+	default 1 if BT_GATT_BAS_LOG_LEVEL_ERR
+	default 2 if BT_GATT_BAS_LOG_LEVEL_WRN
+	default 3 if BT_GATT_BAS_LOG_LEVEL_INF
+	default 4 if BT_GATT_BAS_LOG_LEVEL_DBG
```
Same comment as in the DIS: This is inconsistent with how it is done in all other drivers and subsystems

JoeHut replied:
yeah, I definitely see the advantages, I just don't like that the same functionality is differently defined across the whole tree...

MaureenHelm commented on drivers/sensor/mockup/battery/battery.c:
```diff
@@ -0,0 +1,105 @@
+
+static struct device *device;
+static int battery_init(struct device *dev);
+static sensor_trigger_handler_t trigger_callback;
```
Don't use static variables for the device pointer and trigger_callback, as this will make your driver difficult to instantiate more than once. Move the variables into `struct battery_data` instead and use the `CONTAINER_OF` macro.
https://github.com/zephyrproject-rtos/zephyr/blob/148c7dd8a3fb989e3358425023809f74b71da936/drivers/sensor/fxos8700/fxos8700_trigger.c#L130-L136

JoeHut commented on include/sensor.h:
```diff
@@ -142,6 +142,8 @@ enum sensor_channel {
 	/** Current, in amps **/
 	SENSOR_CHAN_CURRENT,
+	/** State of charge, percent. **/
+	SENSOR_CHAN_SOC,
```
SoC for State of Charge and SoH for State of Health are commonly used in battery fuel gauges

lemrey replied:
It is as far as I know but I don't mind changing this to something different at all

joerchan replied:
I think CHARGE_STATE would be clearer.

lemrey commented on drivers/sensor/mockup/battery/Kconfig:
```diff
@@ -0,0 +1,47 @@
+choice BATTERY_SENSOR_TRIG_POLICY
+	prompt "Trigger policy"
+
+config BATTERY_SENSOR_TRIG_POLICY_PERIODIC
```
Hi @MaureenHelm I only implemented the policy needed to keep the pre-existing functionality in the battery service; I don't plan to implement other policies.

MaureenHelm replied:
Then why make it a choice at all?

lemrey committed:
Bluetooth: services: battery service enhancements

This commit moves the BLE GATT Battery service
from /samples/bluetooth/gatt to /subsys/bluetooth/services and
adds a Kconfig entry to enable and configure the service;
when enabled, it will register itself automatically.

The service now uses the sensor interface to obtain readings
and automatically sends a notification on battey level changes.

Signed-off-by: Emanuele Di Santo <emdi@nordicsemi.no>
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

carlescufi commented:
@JoeHut @MaureenHelm @Vudentz can you revisit this?

carlescufi commented on subsys/bluetooth/services/Kconfig:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if BT_PERIPHERAL
```
Why can't a central have services?

lemrey replied:
Using BT_CONN instead.

JoeHut reviewed (commented):
Despite some nitpicking this looks good to me :+1:

lemrey commented on subsys/bluetooth/services/Kconfig.bas:
```diff
@@ -0,0 +1,39 @@
+	bool "Mockup battery sensor"
+	select MOCKUP_BATTERY_SENSOR
+
+endchoice
```
I see what you mean, however how would I be able to `select` _which_ device driver to compile based on the device name?

JoeHut replied:
Good point actually.

lemrey commented on subsys/bluetooth/services/Kconfig:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+if BT_CONN
```
It shows up fine, personally I'd rather have it in the child Kconfig so that it is self-contained and the parent looks cleaner.

JoeHut replied:
Ok cool :+1:

pdunaj commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,143 @@
+/** @file
```
I would throw this away - this files contains no documentation really

pdunaj commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,143 @@
+
+static int bas_init(struct device *dev)
+{
+	int rc;
```
s/rc/err

pdunaj commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,143 @@
+	if (!battery) {
+		LOG_ERR("Could not bind to device %s",
+			CONFIG_BATTERY_SENSOR_DEV_NAME);
+		return -EIO;
```
`-ENXIO`

lemrey replied:
`-EIO` is fine; the one you suggested might be more appropriate but it is not really used.
`git grep "EIO" | wc -l` : `1019`
`git grep "ENXIO" | wc -l` : `3`

pdunaj commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,143 @@
+	rc = sensor_trigger_set(battery, &trigger, level_changed_callback);
+	if (rc) {
+		LOG_ERR("Could not set up sensor trigger, err %d", rc);
+		return -EIO;
```
`return err;`

pdunaj commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,143 @@
+void level_changed_callback(struct device *dev, struct sensor_trigger *trigger)
+{
+	ARG_UNUSED(dev);
+
```
```
if (!notif_enabled) {
   return;
}

lemrey replied:
That's not what we want; when the sensor notifies a change in battery level the database should be updated regardless of whether notifications are enabled or not.

pdunaj commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,143 @@
+{
+	ARG_UNUSED(dev);
+
+	if (trigger->type == SENSOR_TRIG_DATA_READY &&
```
I would add brackets

jhedberg replied:
No need, actually. This is also consistent with the Bluetooth code base in general.

pdunaj commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,143 @@
+
+static struct bt_gatt_service bas = BT_GATT_SERVICE(attrs);
+
+SYS_INIT(bas_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```
please don't do it.... I would appreciate it application could initialize bas on its own when it is ready

lemrey replied:
This just registers the service in the database, if you have this service that's probably what you want to do. This has been already discussed at lengths in this PR and #9712.

pdunaj commented on subsys/bluetooth/services/bas.c:
```diff
@@ -0,0 +1,143 @@
+
+	ARG_UNUSED(dev);
+
+	battery = device_get_binding(CONFIG_BATTERY_SENSOR_DEV_NAME);
```
This binds the service to only one way of providing battery measurements. Can we keep that separate and simply notify service with a value?

lemrey replied:
Perhaps, but that was not possible to begin with and I won't add that functionality in this PR.

pdunaj replied:
It's rather about removing things from PR...

pdunaj commented:
@pizi-nordic

MaureenHelm commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -41,4 +41,7 @@ add_subdirectory_ifdef(CONFIG_TMP007		tmp007)
 add_subdirectory_ifdef(CONFIG_VL53L0X		vl53l0x)
 
+# Mockup sensor drivers
+add_subdirectory_ifdef(CONFIG_MOCKUP_BATTERY_SENSOR		mockup/battery)
```
Similar to what you did with Kconfig, move this to `drivers/sensor/mockup/CMakeLists.txt`

pizi-nordic reviewed (changes requested):
We should not try to be smarter than application author as we do not know exact requirements he/she is facing.
In my opinion this commit assumes way too much about battery service usage.

Instead we should provide flexible components which might be reused by different applications with different requirements.

Of course, we can add a complete framework to which will integrate all the components needed to achieve given goal but we should not force developer to use it.

Using feature presented in this PR as example (battery measurement with battery service integration) we should provide:

- ADC driver (it is already there).
- Library calculating battery level basing on measurement results (here we can implement smoothing and similar features).
- Battery service.
- Glue which will integrate this together.

In such approach developer will be able reuse parts which are suitable for his/her application and replace ones which are not. Moreover he/she will be able to reuse some data available at given level for other purposes (for example shut down the application if battery level is below threshold).

lemrey commented:
Hi @pizi-nordic , the purpose of this PR was to give the battery service implementation a Kconfig file so that it would no longer be an orphan C file under /samples. I believe the requirements to do so, as expressed by several people in this and other pull requests, are addressed in full in this patch as it is. With these changes, the battery service can be developed further to add new functionality or modify the existing one like support for other data sources, etc. but I do not plan to do that here.

aescolar commented:
@joerchan Are you taking over this PR? if so, feel free to remove the stale PR label.

joerchan commented:
@aescolar Yes, removed the label

@pizi-nordic Can you have a look at lemreys answers to your comments?

pizi-nordic commented:
Hi.
My biggest concern (the reusability of the code) is still not fully addressed (https://github.com/zephyrproject-rtos/zephyr/pull/9735#pullrequestreview-155443188), however I see some progress in good direction.

Could we have an example which shows how to use this service with something else than battery sensor mockup?

aescolar commented:
@joerchan How is this going?

joerchan commented:
I'm closing this PR in favor of #17355, which moves the BAS service out of the samples folder. But does not include adding the Battery sensor.
